### PR TITLE
Move StartupHookerProvider from shared descriptor file

### LINKED ISSUE #49663: StartupHookProvider causing ILLink warnings in default blazor app
Using the latest 6.0 SDK builds:

`dotnet new blazorwasm`
`dotnet publish -c Release -p:SuppressTrimAnalysisWarnings=false`

```
resource ILLink.Substitutions.xml in System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e : warning IL2008: Could not resolve type 'System.StartupHookProvider' [C:\DotNetTest\Net6Blazor\Net6Blazor.csproj]
```

Since `StartupHookProvider` doesn't exist in mono, we can't put its substitution in the 'shared' substitution file:

https://github.com/dotnet/runtime/blob/79ae74f5ca5c8a6fe3a48935e85bd7374959c570/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml#L22-L24

cc @mateoatr

### PR DESCRIPTION
**Pull request by mateoatr**
Should fix #49663.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.xml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<linker>
```
Comment (vitek-karas) [inline src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.xml:1]: Shouldn't this go to https://github.com/dotnet/runtime/blob/main/src/coreclr/System.Private.CoreLib/src/ILLink/ILLinkTrim.xml ?
The libraries folder is shared between the runtimes, so anything runtime specific should ideally go to the runtime specific folder.

File: src/coreclr/System.Private.CoreLib/src/ILLink/ILLink.Suppressions.LibraryBuild.xml [hunk 2]
```diff
@@ -8,5 +8,6 @@
       <property name="Target">M:System.StartupHookProvider.ProcessStartupHooks()</property>
       <property name="Justification">This warning is left in the product so developers get an ILLink warning when trimming an app with System.StartupHookProvider.IsSupported=true.</property>
     </attribute>
+    <linker>
```
Comment (eerhardt) [inline src/coreclr/System.Private.CoreLib/src/ILLink/ILLink.Suppressions.LibraryBuild.xml:11]: ?

File: src/coreclr/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.xml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<linker>
```
Comment (eerhardt) [inline src/coreclr/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.xml:1]: Nothing is including this new file in the project. You need to add a line like:

https://github.com/dotnet/runtime/blob/6791d05494f5aa4d1a789525c2708e775aafdc07/src/mono/System.Private.CoreLib/System.Private.CoreLib.csproj#L136

**Review by eerhardt — APPROVED** [review]
LGTM

**Review by vitek-karas — APPROVED** [review]

**Review by stephentoub — APPROVED** [review]

**Pull request merged**